### ISSUE TITLE
Stryker.NET doesn't handle xUnit v3 properly

### ISSUE THREAD

**Issue by 0xced** [stryker-mutator/stryker-net#3117]
**Describe the bug**

When using xUnit v3, Stryker.NET produces warnings and can't perform proper mutations.

**Reproduction steps**

Checkout the [xunit-v3](https://github.com/serilog-contrib/serilog-formatting-log4net/tree/xunit-v3) branch of the Serilog.Formatting.Log4Net project and run `dotnet tool restore && dotnet stryker` at the root of the repository.

**Expected behavior**

Stryker.NET performs mutations and gives the same score (100%) as when using xUnit v2.

Dashboard with xUnit v2 (on the `main` branch): [100%](https://dashboard.stryker-mutator.io/reports/github.com/serilog-contrib/serilog-formatting-log4net/main)

**Actual behaviour**

Stryker.NET produces many warnings and eventually fails to perform the mutations.

```
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.IndentationSettingsTest.IndentationSettingsToString), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.ThreadIdProperty), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.MessageCDataMode), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.IndentationSettings), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.DomainAndUserNameProperty), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.LogEventLevel), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.MachineNameProperty), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.XmlElementsLineEnding), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.Log4NetTextFormatterTest.Log4JCompatibility), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
[22:03:18 WRN] Runner 2: Initial test run encounter an unexpected test case (Serilog.Formatting.Log4Net.Tests.PublicApi.ApprovePublicApi), mutation tests may be inaccurate. Disable coverage analysis if you have doubts.
```

Dashboard with xUnit v3: [3.09%](https://dashboard.stryker-mutator.io/reports/github.com/serilog-contrib/serilog-formatting-log4net/xunit-v3)

**Desktop**
 - OS: both macOS and Linux (and probably Windows too)
 - Type of project: .NET 6 + 8 + .NET Standard 2.0
 - Framework Version .NET 8
 - Stryker Version: 4.4.1

**Additional context**

I have **not** enabled [Microsoft Testing Platform support in xUnit.net v3](https://xunit.net/docs/getting-started/v3/microsoft-testing-platform) so I don't think this is related to #3094 although I'm not totally sure.

**Comment by rouke-broersma** [comment]
We have quite some custom logic for dealing with specific test framework quirks, I suspect xunit v3 has added yet another layer of quirks.

**Comment by dupdob** [comment]
I predict a big one: it looks like a lot of VsTest design rules are violated here.

**Comment by dupdob** [comment]
## First Analysis
Stryker is not able to run xUnit 3 theories with inline data, that is Stryker detects them correctly but they are skipped by the runner
`TpTrace Warning: 0 : 18292, 12, 2024/11/22, 17:39:20.512, 224768786899, testhost.dll, [xUnit.net 00:00:00.22] Serilog.Formatting.Log4Net.Tests: Skipping test case 'Serilog.Formatting.Log4Net.Tests.IndentationSettingsTest.IndentationSettingsToString' (ID '6066754da9eb486a3555a1ff4792822d9ebe45d5ccf3d02486e8bdb36c7bed24') without serialization
`
it looks like Stryker should provide serialized data and somehow fails to do so.
In theory Stryker keeps all data for everytest and reuse them afterward, but there is probably a bug somewhere.

The good news is that it does not look like a big issue.

**Comment by dupdob** [comment]
## Consequences
1. Stryker does not currently support xUnitV3 test adapter as it deviates too much from the standard behavior
2. The main consequence is that Stryker cannot work **AT ALL** with xUnitV3 test base. Any attempt will result in most mutants being flag as not covered and the remaining flagged as survived.
3. There is no clear path forward to fix this
4. In the short term, Stryker can be modified to support a simpler approach to tests. That is one without coverage analysis and using environment variables for control. Note that it will result in very poor performances.
5. For medium term, some basic form of coverage could be restored (is a mutation covered yes or no), that would only slightly improve performances
6. Alternatively, we could revert to using IPC mechanisms to exchange data between coverage collector and the testing logic. But bear in mind this logic was dropped in the pas as it was complex and a recurrent source of issues.

I do not think trying to fit support for xUnitV3 via VsTest is worth the effort, I suppose the move to MTP will provide a smoother solution. But that remains to be confirmed.

The alternative is to implement a dedicated runner for xUnit V3, which would provide other interesting benefits but looks like a lot of work.

**Comment by dupdob** [comment]
For context:

https://github.com/xunit/xunit/issues/2337

**Comment by rouke-broersma** [comment]
Why have standards.. 🫠

**Comment by dupdob** [comment]
PR #3122 restore basic support for xUnit V3. Basic as no coverage based optimization possible.

**Comment by 0xced** [comment]
Let's ping @bradwilson on this issue. He's been very helpful when I reported issues on the early xUnit v3 prereleases. Maybe he can give some guidance.

**Comment by bradwilson** [comment]
> it looks like there are differences in how theories are captured between xUnit V2 and V3. Main (single?) impact is that theories data set are discovered at run time only

This is controlled by [configuration](https://xunit.net/docs/configuration-files#preEnumerateTheories) and/or [RunSettings](https://xunit.net/docs/runsettings#PreEnumerateTheories). The default for whether we pre-enumerate theories depends on the runner in question: we generally disable it with our console runner, and enable it with `xunit.runner.visualstudio` when we believe we're inside Visual Studio (vs. running from `dotnet test`).

If you are invoking a v3 test assembly in automated mode, then `-preEnumerateTheories` will turn it on (or `--pre-enumerate-theories on` in Microsoft Testing Platform CLI mode).

> It is likely that the xUnit V3 test runner executes the tests outside the VsTest host.

That is correct. v3 test projects are executables. When running v3 tests with `xunit.runner.visualstudio`, it launches those tests into a separate process and does the conversion from the `xunit.v3.runner.utility` object model into the VSTest object model.

> Let's ping @bradwilson on this issue.

Ask away. 😄

**Comment by bradwilson** [comment]
> Can we toggle the https://github.com/xunit/xunit/issues/2947 via some visualstudio runner's custom settings?

No, that ability is only provided to runner authors. You'd need a replacement for xunit.runner.visualstudio that could run things in-process, and you'd be entirely responsible for the dependency resolution. In practice, this isn't something that's easy to do; I personally have tried and failed, so I wouldn't be able to help you do it. Part of the reason that this is a challenging task is that it also involves any unmanaged dependency resolution, assuming you have customers who are using unmanaged dependencies (like C++ libraries).

> this new architecture makes. 'in proc' data collector useless in general.

This isn't just the architecture of xUnit.net v3; it's also the architecture of Microsoft.Testing.Platform. That means MSTest 3.6+ tests which enable MTP will be in the same situation that you find yourself with xUnit.net v3 today.

> For context, Stryker always enables design mode, and we had no problem with the latest xUnit V2.xx.

There may be a bug here, then. When design mode is enabled, pre-enumeration should be on (unless explicitly turned off via config).

**Comment by rouke-broersma** [comment]
@bradwilson we use the vstest translation layer /test platform v2 so currently our design is not compatible with mstest testplatform anyway.

**Comment by MeydanOzeri** [comment]
as of now you can't use xunit3 with stryker.net, is this fix planned anytime soon ?

**Comment by dupdob** [comment]
There is no plan to support xUnit V3 through VSTest for the moment. To be honest, I am not sure it can be done as it requires writing a dedicated xUnit V3 host/runner for that.

So the only way forward is through Ms Testing Platform support

**Comment by candoumbe** [comment]
> There is no plan to support xUnit V3 through VSTest for the moment. To be honest, I am not sure it can be done as it requires writing a dedicated xUnit V3 host/runner for that.
> 
> So the only way forward is through Ms Testing Platform support

That's quite unfortunate 😨 !!!

I rely so heavily on both `Stryker` and `xUnit` and I migrated from xunit-v2 to xunit-v3 without even thinking about consequences on `Stryker` tests.

Could anyone point me to documentation on building host/runner (just to try to better understand what needs to be done)

💪🏾 Keep up the good work guys/gals ...

**Comment by slang25** [comment]
> Could anyone point me to documentation on building host/runner (just to try to better understand what needs to be done)

The problem is there is no "read this, then do that", the solution needs a bespoke design. See the conversation here: https://github.com/stryker-mutator/stryker-net/issues/3094

I'm happy to put some time into this too.

**Comment by bradwilson** [comment]
(Speaking as the xUnit.net person) It's probably worth mentioning that Microsoft Testing Platform v2 will no longer support VSTest on .NET 10. It _probably_ doesn't matter, since the intended approach of supporting MTP in general would be fine, but Microsoft (the TestFx team, at least) is pushing very hard to sunset VSTest and get everybody over onto MTP. All the major test frameworks support it.

**Comment by dupdob** [comment]
The main obstacle we face is that MTP's architecture does not provide any of the extension points needed for a tool such as Stryker. As such, we can only provide basic support for it, meaning no test execution optimizations whatsoever. This implies running time being 10-30 times slower than with VsTest.
We are hoping to have a breakthrough about this and find a way to move forward, but no luck so far.
Edit: this issue was closed by mistake, I reopened it

**Comment by vbreuss** [comment]
@dupdob : I don't understand what the status of this issue is now. Your last comment mentions some obstacles, and it is now closed, without any linked change.
Is this completed as in "at least basic support now works" or rather as in "we can't get it to work"?

**Comment by Frulfump** [comment]
Or should it be closed as not planned or could it be re-opened to track the hope of finding a breakthrough
cc @Youssef1313 if you could provide any insights from Microsofts perspective

**Comment by dupdob** [comment]
Closing the issue was a simple mishap.
Even basic support requires significant work, and I think it will not be satisfactory for anyone except very small projects.

**Comment by Youssef1313** [comment]
I haven't went through all comments, but the general main issue I'm seeing is the ability to run Stryker VSTest extension in the same process that runs the tests. This simply cannot be done AFAIK with the current design of xunit.v3. **However**, this **can** be done for xunit.v3 when it's run via Microsoft.Testing.Platform. I would expect most xunit.v3 users to be using MTP. So what might be needed here is investigating the support for MTP in general. But I don't have much context on how Stryker works internally.

**Comment by slang25** [comment]
@Youssef1313 I'm not a member of the stryker team, however I'm happy to jump on a call and explain how stryker current works, how it interacts with vstest, what optimizations it makes etc...

If that's helpful then let me know and I'll email you to continue the conversation.

**Comment by dupdob** [comment]
In theory, (source level) mutation testing only needs the following capabilities:
1) being able to inject mutations in the project source code
2) being able to build said project
3) tests each mutation against the test base.
But the naive approach leads to hours long test sessions due to the test base being run for hundreds or even thousands of mutations. So most mutation tools perform some form of coverage analysis.

In its simplest form, coverage analysis identifies mutants that are not covered by any test and flag those as survived without having to test them. Alas, this brings only marginal gains on project with a decent test coverage. And it turns out most teams interested in mutation testing do have a good test coverage.
Non trivial coverage consists of identifying which tests cover each mutant and then only run required tests when testing said mutant. This approach leads to huge gain on medium to large code bases, like 10-20 times faster.

Stryker goes even one step further by testing several mutants in a single test sessions. It does that selecting mutants that have distinct coverage, ensuring each run test only test for one mutant.

In order to capture test coverage, Stryker needs to monitor what happens during each test. Coverage capture needs to happen in process and needs to be associated with the test id. For that to be possible, Stryker currently uses event raised by the in-process extension architecture provided by VsTest.
As xUnit V3 runs out of process, coverage data can no longer be accessed by the extension. Furthermore, events are raised asynchronously, meaning coverage data can no longer be reliably associated with the test id.

MTPs put us in worse situation as it is fully async.

As for as I understand it, the only way we could **theorically** makes Stryker works with MTP is by creating **test framework specific extensions** to help us capture the coverage. Which means
1. an awful lot of code to write
2. understanding the architecture of said test frameworks
3. maintaining said extensions according the test framework releases
-------
None of this is sustainable for a free OSS project

**Comment by bradwilson** [comment]
> I would expect most xunit.v3 users to be using MTP.

We don't think of it this way.

xUnit.net v3 _always_ supports MTP, even if you aren't always using it. I expect users to be "sometimes using MTP, and sometimes not", and it really depends on the user. Our default project template includes a package reference to `xunit.runner.visualstudio` which is what adds VSTest support.

Some hopefully clarifying thoughts here:

- In xUnit.net v3, support for VSTest is _**orthogonal**_ to support for MTP. You can disable support for VSTest by removing the `xunit.runner.visualstudio` reference, but you cannot disable support for MTP.

- xUnit.net v3 test projects always run in a separate process, whether that's run via our own native runner APIs (i.e., what's used by `xunit.v3.runner.console` and `xunit.v3.runner.msbuild`), or if that's via VSTest, or if that's via MTP. The difference lies in where the adapter-to-native code lives: for VSTest, it's in the VSTest process (via code in `xunit.runner.visualstudio`); for MTP, it's in the test project process (via code that is always compiled into xUnit.net v3 test projects). This is what makes a VSTest-based adapter unworkable for your situation, but an MTP-based adapter (potentially) workable.

- MTP can optionally be used for `dotnet test`. This depends on whether the user has defined `TestingPlatformDotnetTestSupport` to be `true` in the project `PropertyGroup`. By default, it uses VSTest.

- MTP will be used by default for Test Explorer in current builds of VS, with the ability to disable that by setting `DisableTestingPlatformServerCapability` to `true`. Older versions of Test Explorer either only support VSTest or required a preview feature flag to enable MTP support.

- Third party test runners may be based on MTP or based on VSTest, and xUnit.net v3 can support both. In rare cases, third party test runners may also be based on our native runner APIs.

- When running a test project directly (i.e., via `dotnet run`), it defaults to an xUnit.net native command line UX experience. You can override this by setting `UseMicrosoftTestingPlatformRunner` to `true`, though that only affects the command line experience when running the test project directly. It does not affect or interfere with the ability to use any of the other execution methods listed above (for example, you can enable the command line MTP experience but still be using VSTest via `dotnet test`).

This is why a statement like "I would expect most `xunit.v3` users to be using MTP" is not really a statement you can make. It's rarely just one or the other, when there are at least five different ways to run xUnit.net v3 tests: directly (xUnit.net native UX), directly (MTP UX), via xUnit.net native runner APIs, via VSTest, and via MTP.

**Comment by bradwilson** [comment]
Also, about half of what I wrote above changes with .NET 10 SDK coming in November.

**Comment by Youssef1313** [comment]
> MTP can optionally be used for dotnet test. This depends on whether the user has defined TestingPlatformDotnetTestSupport to be true in the project PropertyGroup. By default, it uses VSTest.

Yeah, and that's definitely changing in .NET 10. From pure MTP perspective, if `IsTestingPlatformApplication` is true (which is the case by default for xunit.v3), then it's an MTP test project. So, xunit.v3 projects have always been MTP. The thing is that we had an extra gate to opt-in to using MTP in the VSTest-based `dotnet test`, which is `TestingPlatformDotnetTestSupport`. But once that gets out of the picture, and we force MTP projects in .NET 10 SDK to use the new `dotnet test` experience, that means that we view xunit.v3 projects as MTP projects (even if they support VSTest and reference xunit.runner.visualstudio and/or Microsoft.NET.Test.Sdk).

> When running a test project directly (i.e., via dotnet run), it defaults to an xUnit.net native command line UX experience. You can override this by setting UseMicrosoftTestingPlatformRunner to true, though that only affects the command line experience when running the test project directly. It does not affect or interfere with the ability to use any of the other execution methods listed above (for example, you can enable the command line MTP experience but still be using VSTest via dotnet test).

Yeah I see how `UseMicrosoftTestingPlatformRunner` works (just controls the entry point, and both entrypoints support MTP when run with `--server` or `--internal-msbuild-node`). That's very different from how other test frameworks (Expecto, MSTest, and NUnit) does it, which creates user confusion IMO. While the documentation I read today on your side looks pretty accurate, I have doubts it's well understood by users.

**Comment by bradwilson** [comment]
> That's very different from how other test frameworks (Expecto, MSTest, and NUnit) does it, which creates user confusion IMO.

If every test framework were the same, then we would only need one test framework. 😄 Is it confusing that TUnit doesn't support VSTest, and thus considered some kind of deficiency because it's "different"?

**Comment by tknightnd** [comment]
Hi,

Unfortunately [XUnit v2.9.3 has been deprecated](https://www.nuget.org/packages/xunit/2.9.3). The problem for projects in regulated environments is that they need upgrading away as soon as possible, so back peddling to v2.9.2 simply isn't an option.

Clearly there's a lot of nasty work to be done to solve this issue, but is there any traction?

**Comment by slang25** [comment]
@tknightnd there sure is, @richardwerkman has been working on a branch which adds this support and it's looking promising 🤞

**Comment by bradwilson** [comment]
@tknightnd We will continue to support releasing security fixes for v2, if ever one came in. In the nearly 20 years of xUnit.net, this has never happened. Similarly, if a showstopping bug were discovered with no workaround, we would issue a new build with a fix (but my belief that this might happen is near zero at this point, given the age and stability of the v2 code base).

The word "done" might be better than "deprecated", but that's not an option on NuGet. 😄 What it really means is "there are no more features coming here, all focus is on v3".

**Comment by richardwerkman** [comment]
@tknightnd see #3094 for the status on adding support for xUnit v3. I'll give updates there.

**Comment by tknightnd** [comment]
@richardwerkman @bradwilson @slang25  - thank you all for the positive, prompt answers - you're all legends.
